include/fmt/core.h

```diff
@@ -2244,21 +2255,87 @@ class compile_parse_context
   using base::check_arg_id;
 };
 
+template <typename T> constexpr bool check_named_argument() {
+  return !detail::is_named_arg<T>() ||
+         (detail::is_named_arg<T>() && detail::is_statically_named_arg<T>());
+}
+
+template <typename... Args, FMT_ENABLE_IF(sizeof...(Args) == 0)>
+constexpr bool check_named_arguments() {
+  return true;
+}
+
+template <typename T, typename... Args> constexpr bool check_named_arguments() {
+  return check_named_argument<T>() && check_named_arguments<Args...>();
+}
+
+constexpr int invalid_arg_index = -1;
+
+#if FMT_USE_NONTYPE_TEMPLATE_PARAMETERS
+template <int N, typename T, typename... Args, typename Char>
+constexpr int get_arg_index_by_name(basic_string_view<Char> name) {
+  if constexpr (detail::is_statically_named_arg<T>()) {
+    if (name == T::name) return N;
+  }
+  if constexpr (sizeof...(Args) == 0) {
+    return invalid_arg_index;
+  } else {
+    return get_arg_index_by_name<N + 1, Args...>(name);
+  }
+}
+
+template <typename... Args, typename Char>
+constexpr int get_arg_index_by_name(basic_string_view<Char> name) {
+  if constexpr (sizeof...(Args) == 0) {
+    return invalid_arg_index;
+  } else {
+    return get_arg_index_by_name<0, Args...>(name);
+  }
+}
+#else
+template <typename... Args, typename Char>
+constexpr int get_arg_index_by_name(basic_string_view<Char>) {
+  return invalid_arg_index;
+}
+#endif
+
 template <typename Char, typename ErrorHandler, typename... Args>
 class format_string_checker {
  public:
   explicit FMT_CONSTEXPR format_string_checker(
       basic_string_view<Char> format_str, ErrorHandler eh)
       : context_(format_str, num_args, eh),
-        parse_funcs_{&parse_format_specs<Args, parse_context_type>...} {}
+        parse_funcs_{&parse_format_specs<Args, parse_context_type>...} {
+    static_assert(
+        check_named_arguments<Args...>(),
+#if FMT_USE_NONTYPE_TEMPLATE_PARAMETERS
+        "use statically named arguments with compile-time format strings: "
+        "fmt::arg(\"name\", value) -> \"name\"_a = value"
+#else
+        "compile-time checks don't support named arguments, "
+        "please use C++20 or/and a newer compiler"
+#endif
+    );
+  }
 
   FMT_CONSTEXPR void on_text(const Char*, const Char*) {}
 
   FMT_CONSTEXPR int on_arg_id() { return context_.next_arg_id(); }
   FMT_CONSTEXPR int on_arg_id(int id) { return context_.check_arg_id(id), id; }
-  FMT_CONSTEXPR int on_arg_id(basic_string_view<Char>) {
-    on_error("compile-time checks don't support named arguments");
+  FMT_CONSTEXPR int on_arg_id(basic_string_view<Char> id) {
+#if FMT_USE_NONTYPE_TEMPLATE_PARAMETERS
+    auto index = get_arg_index_by_name<Args...>(id);
+    if (index == invalid_arg_index) {
+      on_error("named argument is not found");
+    }
+    return context_.check_arg_id(index), index;
+#else
+    (void)id;
+    on_error(
+        "compile-time checks don't support named arguments, "
+        "please use C++20 or/and a newer compiler");
```

Comment:
I suggest something like "compile-time checks for named arguments require C++20 support".